### ISSUE TITLE
not working

### ISSUE THREAD

**Issue by huahuayu** [magnetikonline/markdown-toc-generate#2]
hi, I tried your toc generate online, I have a bug report, it's not working, some of `##` title missing(say 查看运行的容器列表) and the link is not right, please check.
my markdown is as below:
https://github.com/huahuayu/yum-docker-cookbook/blob/master/README.md

**Comment by magnetikonline** [comment]
Thanks @huahuayu - can you determine the first heading that's failing? As the README.md is quite long - if you can isolate that it would help.

I assume it's the UTF-8 headings causing the issue, something I hadn't thought of!

Cheers!

**Edit:** actually, looking at the `README.md` - it seems the first header that fails is `docker命令自动补全`? Is that correct?

**Comment by huahuayu** [comment]
hi，for [above](https://github.com/huahuayu/yum-docker-cookbook/blob/master/README.md) markdown, now your toc output is  
```
- [yum docker cookbook <working in progress>](#yum-docker-cookbook-working-in-progress)
	- [查看版本](#)
	- [查看帮助](#-1)
	- [命令格式](#-2)
	- [hello docker](#hello-docker)
	- [指定容器名字](#-3)
	- [删除容器](#-4)
	- [容器和虚拟机的区别](#-5)
	- [docker create/start/run区别](#docker-createstartrun)
	- [查看容器元数据](#-6)
	- [运行容器的同时启动bash](#bash)
	- [在启动的容器中执行命令](#-7)
	- [在容器中中运行alpine](#alpine)
	- [进入容器的方法总结](#-8)
	- [查看容器端口](#-9)
	- [查看容器的ip地址](#ip)
	- [容器网络](#-10)
	- [新建网络](#-11)
	- [指定容器网络](#-12)
	- [容器链接网络](#-13)
	- [断开网络](#-14)
```

I don't know what's the #-1,-2,-3 means，but I know many "##" title is missing. If I take the () off, it's as below:  
``` 
- [yum docker cookbook <working in progress>]
- [查看版本]
- [查看帮助]
- [命令格式]
- [hello docker]
- [指定容器名字]
- [删除容器]
- [容器和虚拟机的区别]
- [docker create/start/run区别]
- [查看容器元数据]
- [运行容器的同时启动bash]
- [在启动的容器中执行命令]
- [在容器中中运行alpine]
- [进入容器的方法总结]
- [查看容器端口]
- [查看容器的ip地址]
- [容器网络]
- [新建网络]
- [指定容器网络]
- [容器链接网络]
- [断开网络]
```

the correct sequence should be  
```
- [yum docker cookbook <working in progress>]
- [查看版本]
- [查看参数]
- [查看帮助]
- [查看特定命令帮助]
- [命令格式]
- [hello docker]
- [docker命令自动补全]
- [查看运行的容器列表]
- [停止容器运行]
- [指定容器名字]
- [给容器重命名]
- [查看日志]
- [删除容器]
- [docker run命令详解]
- [查看容器中的进程]
- [容器和虚拟机的区别]
- [docker create/start/run区别]
- [查看容器元数据]
- [设置环境变量]
- [查看所有容器运行状态]
- [运行容器的同时启动bash]
- [在容器中运行ubuntu]
- [在启动的容器中执行命令]
- [将image拉到本地]
- [查看image列表]
- [在容器中中运行alpine]
- [进入容器的方法总结]
- [查看容器端口]
- [查看容器的ip地址]
- [容器网络]
- [查看网络列表]
- [默认网络]
- [查看网络详情]
- [新建网络]
- [指定容器网络]
- [容器链接网络]
- [断开网络]
- [教程推荐]
```

you don't know Chinese characters doesn't matter, now they are comparable, this is the compare result I got from https://text-compare.com/
![image](https://user-images.githubusercontent.com/1725060/53814374-7875a480-3f9a-11e9-8d7c-1369e8a8c154.png)

**Comment by magnetikonline** [comment]
Thanks for the detailed writeup! I'll take a look.

**Comment by magnetikonline** [comment]
Hey @huahuayu I haven't forgotten about this!

Found some progress today - leaving this here, so I don't forget.

https://github.com/github/cmark-gfm/blob/aed182ed089f1c4d42b75657064ae76904e9e024/src/utf8.c#L256-L317

Will be in touch...

**Comment by huahuayu** [comment]
Great!

**Comment by magnetikonline** [comment]
Hello @huahuayu I think we're done!!

This commit 315bbda8bf921bab06a4feb16289e43d4d46c225 - which also includes tests.

Implemented a version of CommonMark's (used by GitHub's Markdown render) `int cmark_utf8proc_is_punctuation(int32_t uc)` - which as the name suggests determines if a character is punctuation - vital in doing this right for extended UTF-8 characters (such as yours).

I've tested your `README.md` at the start of this issue (https://github.com/huahuayu/yum-docker-cookbook/blob/master/README.md) in a dummy Gist - and can reach all headings.

I think this makes this one of the better TOC generator implementations, so many I found that complain of this same issue you've raised - so pretty happy I've been able to (hopefully!) solve for it!

Let me know how you go. If all good - we can close this out! 🎉
